src/jit/codegenxarch.cpp

```diff
@@ -3246,51 +3245,40 @@ CodeGen::genLclHeap(GenTreePtr tree)
 
         // else, "mov regCnt, amount"
 
-#if defined(_TARGET_X86_)
         if (compiler->info.compInitMem)
         {
-            // For x86, when initializing memory with a constant count, we want 'amount' to be the
-            // count of pointer-sized words, not bytes.
-            amount = cntPtrSizedWords;
+            // When initializing memory, we want 'amount' to be the loop count.
+            assert((amount % STACK_ALIGN) == 0);
+            amount /= STACK_ALIGN;
         }
-#endif // _TARGET_X86_
 
         genSetRegToIcon(regCnt, amount, ((int)amount == amount)? TYP_INT : TYP_LONG);
     }
 
     loop = genCreateTempLabel();
     if (compiler->info.compInitMem)
     {
-        // At this point 'regCnt' is set to the total number of bytes (or words, for constant counts) to locAlloc.
+        // At this point 'regCnt' is set to the number of loop iterations for this loop, if each
+        // iteration zeros (and subtracts from the stack pointer) STACK_ALIGN bytes.
         // Since we have to zero out the allocated memory AND ensure that RSP is always valid
         // by tickling the pages, we will just push 0's on the stack.
-        // 
-        // Note: regCnt is guaranteed to be even on Amd64 since STACK_ALIGN/TARGET_POINTER_SIZE = 2
-        // and localloc size is a multiple of STACK_ALIGN.
 
         assert(genIsValidIntReg(regCnt));
 
         // Loop:
         genDefineTempLabel(loop);
 
 #if defined(_TARGET_AMD64_)
-        // dec is a 2 byte instruction, but sub is 4 (could be 3 if
-        // we know size is TYP_INT instead of TYP_I_IMPL)
-        // Also we know that we can only push 8 bytes at a time, but
-        // alignment is 16 bytes, so we can push twice and do a sub
-        // for just a little bit of loop unrolling
+        static_assert_no_msg(STACK_ALIGN == (REGSIZE_BYTES * 2));
```

Comment:
I realize that this would be self-evident and redundant with what is above, but a small comment noting that this relies on the fact that STACK_ALIGN is 16 bytes would be useful.